catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/DurableFileAlterationListener.java

```diff
@@ -0,0 +1,115 @@
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javafx.util.Pair;
+import javax.validation.constraints.NotNull;
+import org.apache.commons.io.monitor.FileAlterationListenerAdaptor;
+import org.codice.ddf.platform.util.StandardThreadFactoryBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DurableFileAlterationListener extends FileAlterationListenerAdaptor {
+
+  public static final long DEFAULT_PERIOD = 5;
+
+  public static final String CDM_FILE_CHECK_PERIOD_PROPERTY = "org.codice.ddf.cdm.fileCheckPeriod";
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DurableFileAlterationListener.class);
+
+  private Map<File, Pair<Long, WatchEvent.Kind<Path>>> fileMap = new ConcurrentHashMap<>();
+
+  private ScheduledExecutorService executorService;
+
+  private AbstractDurableFileConsumer consumer;
+
+  public DurableFileAlterationListener(@NotNull AbstractDurableFileConsumer consumer) {
+    this.consumer = consumer;
+    executorService =
+        Executors.newSingleThreadScheduledExecutor(
+            StandardThreadFactoryBuilder.newThreadFactory("directoryMonitorFileChecker"));
+    this.startExecutor();
+  }
+
+  public DurableFileAlterationListener(
+      @NotNull AbstractDurableFileConsumer consumer,
+      @NotNull ScheduledExecutorService executorService) {
+    this.consumer = consumer;
+    this.executorService = executorService;
+    this.startExecutor();
+  }
+
+  private void startExecutor() {
+    long period = DEFAULT_PERIOD;
+    try {
+      period =
+          Long.parseLong(
+              System.getProperty(CDM_FILE_CHECK_PERIOD_PROPERTY, Long.toString(DEFAULT_PERIOD)));
+      if (period < 1) {
+        period = DEFAULT_PERIOD;
+      }
+    } catch (NumberFormatException e) {
+      LOGGER.warn(
+          "Invalid value for system property org.codice.ddf.cdm.fileCheckPeriod. Expected an integer but was {}. Defaulting to {}",
+          System.getProperty(CDM_FILE_CHECK_PERIOD_PROPERTY),
+          period);
+    }
+    executorService.scheduleAtFixedRate(this::checkFiles, 10, period, TimeUnit.SECONDS);
+  }
+
+  public void destroy() {
+    executorService.shutdownNow();
+  }
+
+  @Override
+  public void onFileChange(File file) {
+    if (!fileMap.containsKey(file)) {
+      fileMap.put(file, new Pair<>(-1L, StandardWatchEventKinds.ENTRY_MODIFY));
+    }
+  }
+
+  @Override
+  public void onFileCreate(File file) {
+    fileMap.put(file, new Pair<>(-1L, StandardWatchEventKinds.ENTRY_CREATE));
+  }
+
+  @Override
+  public void onFileDelete(File file) {
+    consumer.createExchangeHelper(file, StandardWatchEventKinds.ENTRY_DELETE);
+  }
+
+  void checkFiles() {
```

Comment:
:question: Seems like this could still be a problem when files are copied slowly.  It's too bad that we can't have the code that is doing to copying of the file into the CDM directory name the file with a different extension until it is fully copied.  This way knowing when the file was completely copied would be under the control of the thing actually writing the file.

Reply:
Unfortunately we have no control over that code, as it could be anything writing to that directory. There are some conventions we could additionally check for, but those are implementation-dependent, so we would still need a fall-back logic.